lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -568,8 +572,9 @@ SmallVector<Value> lowerLdSt(
           reps.apply({{kReg, j}, {kLane, 0}, {kWarp, 0}})[0].second;
       auto regIdxAddI8 = regIdxAdd * (bitwidth / 8);
       Value innerOffset = b.add(offset, b.i32_val(regIdxAddI8));
-      auto vecAddr = b.gep(smemPtrTy, i8_ty, smemBase, innerOffset,
-                           LLVM::GEPNoWrapFlags::inbounds);
+      auto vecAddr =
+          b.gep(smemPtrTy, i8_ty, smemBase, smemAddrAddon(innerOffset),
```

Comment:
Should this be always on actually rather than an addon?

Reply:
If there is no padding, this function is just an identity function.

Reply:
Oh I meant whether we can do the check of whether padding or not directly in `lowerLdSt` and not using a lambda to plug it in. Taking a closer look it would need the source encoding type which is not passed into `lowerLdSt` so this is probably fine then.

Reply:
Actually this is sub-optimal if we have multi-layer of padding. I tried to apply padded offset separately on the Value and int parts of the smemOffset and it works fine. Here is the draft commit bb19837d979ba844feb9163063421d4a619ce53d. We will open a PR with this optimization for in-thread-transpose + paddedShared layout.

Reply:
I don't think that splitting the value is correct in general. that function is not linear wrt. xor (the addition is not) so you cannot split it.
It could be that you could apply that optimisation selectively depending on the layout for some layouts, but I don't think you can do it in general.